### ISSUE TITLE
LISP; SETF FASL

### ISSUE THREAD

**Issue by eswenson1** [PDP-10/its#1275]
Currently, the DB built doesn't build SETF FASL from source.  The reason, up to now, has been that it didn't compile.  The latest source we have for this is NILCOM; SETF 293.  I don't believe there ever was a later version of MacLISP.  Attempting to compile it gets the error (there actually is a second, similar error):

```
(COMMENT **ERROR**  (EVAL-ORDERED* (LIST* (QUOTE A) (APPEND GENSYMS (QUOTE (V)))) (LIST* ARRAY (APPEND INDICES (LIST VAL))) (LIST !
(QUOTE LIST) (LIST* (QUOTE LIST) (QUOTE (QUOTE STORE)) (LIST* (QUOTE LIST) (QUOTE (QUOTE ARRAYCALL)) (LIST (QUOTE QUOTE) (SETF-USE!
R-SLOT EXPF)) (QUOTE A) GENSYMS) (QUOTE (V)))))
                Wrong number of args in function |ARRAYCALL SETF-X-INVERT|)
; DATA ERROR - TO PROCEED TYPE $P
;BKPT DATA
```

What the compiler is saying is that the call to eval-ordered* is supplying an incorrect number of paramers when called from |ARRAYCALL SETF-X-INVERT|.  Now, this frequently occurs because the function in question takes a variable number of arguments, but this fact isn't known to the compiler. What typically happens is that the compiler examines the first call to an function it doesn't know anything about and remembers the number of arguments it was passed.  As long as all other calls in the compilation unit pass the same number of arguments to that function, all is ok.  However, if the compiler sees another call with a different number of parameters, the above "Wrong number of arguments in function..." error is raised.

However, in this case, there are two calls to eval-ordered* in NILCOM; SETF 293; and both calls pass 3 arguments.  So that is not the issue here.

It turns out the issue was caused by the fact that SETF, itself, declares eval-ordered* (which is actually defined as an EXPR/SUBR in LISP; EVONCE FASL) as an LEXPR.  I have no idea why it does that.  And, since the compiler loads SETF and load/dump time, the compiler has recorded that eval-ordered* is an LEXPR.  To fix everything, you have to do a few things:

1) update SETF to not do a (*LEXPR EVAL-ORDERED*).
2) bring up COMPLR, break to a REPL, and remove the property with (REMPROP 'EVAL-ORDERED* '*LEXPR)
3) then compile the updated SETF, installing the resulting FASL as LISP; SETF FASL
4) re-load/dump the compiler so that it loads the new SETF

The resulting compiler will not think eval-ordered* is an LEXPR and subsequently dumped compilers won't either.  

Now, this issue leads me to three questions:

1) Why did SETF include this code?:

```
#M (eval-when (eval load compile)
     (and (status feature COMPLR)
         (*lexpr EVAL-ORDERED* SETF-STRUCT))
     )
```

2) what harm will I cause by removing this declaration (when compiling)?

3) Should I remove this for SETF-STRUCT too?  SETF-STRUCT (also defined in SETF) is also a fixed-argument EXPR/SUBR.

**Comment by eswenson1** [comment]
It is conceivable that EVAL-ORDERED* used to take a variable number of arguments, or used &rest to specify the body.  In that case, compiling code that called it *would* require the *LEXPR declaration.  Then, someone changed EVONCE, and didn't update the offending code in SETF.  I'll try to find other versions of EVONCE to check.

**Comment by eswenson1** [comment]
I think my explanation above is probably the right one.  I tried the following.  I created a file (EJS; BAZ >) that contained:
```
(defun bar (a b c) (foo a b c))
(defun baz (a b c d) (foo a b c d))
```

I tried to compile that and it failed with a wrong-type-argument.  My unknown-to-the-compiler definition of foo is:

```
(defun foo (a b &rest c) nil)
```

So the above calls are perfectly legal.  

Adding the following above the definitions of bar and baz:
```
(eval-when (compile)
        (*lexpr foo))
```
Causes it to compile fine and the resulting FASL to load/run fine.
